File: src/queenscoach/realtime.py
"""GTFS-Realtime feed access: vehicle positions, trip updates, service alerts.

Feed quirks observed in the live CATS data, which the shapes below reflect:

- ``VehiclePosition.stop_id`` and ``current_stop_sequence`` do not correspond
  to any stop or sequence in the published static schedule (0% of stop ids
  resolved, and sequences exceed the trip's own stop count). They are therefore
  never surfaced as stop references; next-stop information comes from the
  TripUpdates feed, whose stop ids resolve completely.
- ``StopTimeEvent.delay`` is never populated, but ``time`` and
  ``scheduled_time`` both are, so schedule deviation is computed from those.
"""

from __future__ import annotations

import math
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, Protocol, TypeVar

from google.protobuf.message import DecodeError
from google.transit import gtfs_realtime_pb2 as gtfs

from .cache import Cached, TtlCache
from .config import Config
from .feed_http import fetch_binary


class RealtimeDecodeError(Exception):
    """A realtime feed was not a decodable GTFS-Realtime message."""


@dataclass(frozen=True, slots=True)
class VehiclePosition:
    entity_id: str
    #: Agency vehicle number as shown on the bus or train, e.g. ``2301``.
    vehicle_label: str | None
    vehicle_id: str | None
    trip_id: str | None
    route_id: str | None
    latitude: float
    longitude: float
    bearing_degrees: float | None
    speed_meters_per_second: float | None
    occupancy_status: str | None
    current_status: str | None
    #: Unix time, in seconds, the position was reported.
    timestamp: int | None


@dataclass(frozen=True, slots=True)
class StopTimeUpdate:
    stop_id: str | None
    stop_sequence: int | None
    #: Predicted arrival, Unix time in seconds.
    arrival_time: int | None
    scheduled_arrival_time: int | None
    departure_time: int | None
    schedule_relationship: str | None


@dataclass(frozen=True, slots=True)
class TripUpdate:
    trip_id: str | None
    route_id: str | None
    vehicle_label: str | None
    vehicle_id: str | None
    stop_time_updates: tuple[StopTimeUpdate, ...]


@dataclass(frozen=True, slots=True)
class ServiceAlert:
    header_text: str | None
    description_text: str | None
    cause: str | None
    effect: str | None
    severity_level: str | None
    informed_route_ids: tuple[str, ...]
    informed_stop_ids: tuple[str, ...]


def _optional_str(message: Any, name: str) -> str | None:
    """Read a proto2 optional string, distinguishing unset from empty."""
    return getattr(message, name) if message.HasField(name) else None


def _optional_int(message: Any, name: str) -> int | None:
    return int(getattr(message, name)) if message.HasField(name) else None


def _optional_float(message: Any, name: str) -> float | None:
    if not message.HasField(name):
        return None
    value = float(getattr(message, name))
    return value if math.isfinite(value) else None


def _enum_name(enum_type: Any, message: Any, name: str) -> str | None:
    if not message.HasField(name):
        return None
    try:
        return str(enum_type.Name(getattr(message, name)))
    except ValueError:
        # An unrecognized value is feed data we do not model; report it as absent.
        return None


def _translated_text(translated: Any) -> str | None:
    """Pick the English translation of a GTFS-RT ``TranslatedString``."""
    translations = list(translated.translation)
    if not translations:
        return None
    chosen = next(
        (entry for entry in translations if entry.language.startswith("en")), translations[0]
    )
    return chosen.text or None


def _decode_feed(data: bytes) -> gtfs.FeedMessage:
    """Parse a GTFS-Realtime protobuf.

    Raises:
        RealtimeDecodeError: if the payload is not a valid FeedMessage.
    """
    feed = gtfs.FeedMessage()
    try:
        feed.ParseFromString(data)
    except (DecodeError, ValueError) as error:
        raise RealtimeDecodeError("Realtime feed was not a decodable protobuf") from error
    return feed


def decode_vehicle_positions(data: bytes) -> list[VehiclePosition]:
    """Decode the VehiclePositions feed, dropping entries without coordinates."""
    vehicles: list[VehiclePosition] = []
    for entity in _decode_feed(data).entity:
        if not entity.HasField("vehicle"):
            continue
        vehicle = entity.vehicle

        position = vehicle.position
        latitude = _optional_float(position, "latitude") if vehicle.HasField("position") else None
        longitude = _optional_float(position, "longitude") if vehicle.HasField("position") else None
        # Without coordinates the entry cannot answer a location question.
        if latitude is None or longitude is None:
            continue

        trip = vehicle.trip
        descriptor = vehicle.vehicle
        vehicles.append(
            VehiclePosition(
                entity_id=entity.id,
                vehicle_label=_optional_str(descriptor, "label"),
                vehicle_id=_optional_str(descriptor, "id"),
                trip_id=_optional_str(trip, "trip_id"),
                route_id=_optional_str(trip, "route_id"),
                latitude=latitude,
                longitude=longitude,
                bearing_degrees=_optional_float(position, "bearing"),
                speed_meters_per_second=_optional_float(position, "speed"),
                occupancy_status=_enum_name(
                    gtfs.VehiclePosition.OccupancyStatus, vehicle, "occupancy_status"
                ),
                current_status=_enum_name(
                    gtfs.VehiclePosition.VehicleStopStatus, vehicle, "current_status"
                ),
                timestamp=_optional_int(vehicle, "timestamp"),
            )
        )
    return vehicles


def decode_trip_updates(data: bytes) -> list[TripUpdate]:
    """Decode the TripUpdates feed."""
    updates: list[TripUpdate] = []
    for entity in _decode_feed(data).entity:
        if not entity.HasField("trip_update"):
            continue
        update = entity.trip_update

        stop_times = tuple(
            StopTimeUpdate(
                stop_id=_optional_str(stop_time, "stop_id"),
                stop_sequence=_optional_int(stop_time, "stop_sequence"),
                arrival_time=_optional_int(stop_time.arrival, "time")
                if stop_time.HasField("arrival")
                else None,
                scheduled_arrival_time=_optional_int(stop_time.arrival, "scheduled_time")
                if stop_time.HasField("arrival")
                else None,
                departure_time=_optional_int(stop_time.departure, "time")
                if stop_time.HasField("departure")
                else None,
                schedule_relationship=_enum_name(
                    gtfs.TripUpdate.StopTimeUpdate.ScheduleRelationship,
                    stop_time,
                    "schedule_relationship",
                ),
            )
            for stop_time in update.stop_time_update
        )

        updates.append(
            TripUpdate(
                trip_id=_optional_str(update.trip, "trip_id"),
                route_id=_optional_str(update.trip, "route_id"),
                vehicle_label=_optional_str(update.vehicle, "label")
                if update.HasField("vehicle")
                else None,
                vehicle_id=_optional_str(update.vehicle, "id")
                if update.HasField("vehicle")
                else None,
                stop_time_updates=stop_times,
            )
        )
    return updates


def decode_alerts(data: bytes) -> list[ServiceAlert]:
    """Decode the Alerts feed."""
    alerts: list[ServiceAlert] = []
    for entity in _decode_feed(data).entity:
        if not entity.HasField("alert"):
            continue
        alert = entity.alert

        route_ids: dict[str, None] = {}
        stop_ids: dict[str, None] = {}
        for selector in alert.informed_entity:
            if selector.HasField("route_id"):
                route_ids[selector.route_id] = None
            if selector.HasField("stop_id"):
                stop_ids[selector.stop_id] = None

        alerts.append(
            ServiceAlert(
                header_text=_translated_text(alert.header_text),
                description_text=_translated_text(alert.description_text),
                cause=_enum_name(gtfs.Alert.Cause, alert, "cause"),
                effect=_enum_name(gtfs.Alert.Effect, alert, "effect"),
                severity_level=_enum_name(gtfs.Alert.SeverityLevel, alert, "severity_level"),
                informed_route_ids=tuple(route_ids),
                informed_stop_ids=tuple(stop_ids),
            )
        )
    return alerts


class RealtimeFeeds(Protocol):
    """The three realtime feeds, each cached independently."""

    def vehicle_positions(self) -> Awaitable[Cached[list[VehiclePosition]]]: ...

    def trip_updates(self) -> Awaitable[Cached[list[TripUpdate]]]: ...

    def alerts(self) -> Awaitable[Cached[list[ServiceAlert]]]: ...


@dataclass(frozen=True, slots=True)
class CachedRealtimeFeeds:
    """Feeds backed by TTL caches over the configured feed URLs."""

    _vehicles: TtlCache[list[VehiclePosition]]
    _trips: TtlCache[list[TripUpdate]]
    _alerts: TtlCache[list[ServiceAlert]]

    def vehicle_positions(self) -> Awaitable[Cached[list[VehiclePosition]]]:
        return self._vehicles.get()

    def trip_updates(self) -> Awaitable[Cached[list[TripUpdate]]]:
        return self._trips.get()

    def alerts(self) -> Awaitable[Cached[list[ServiceAlert]]]:
        return self._alerts.get()


_FeedT = TypeVar("_FeedT")


def _feed_cache(config: Config, url: str, decode: Callable[[bytes], _FeedT]) -> TtlCache[_FeedT]:
    async def load() -> _FeedT:
        data = await fetch_binary(
            url,
            timeout_seconds=config.request_timeout_seconds,
            max_bytes=config.max_feed_bytes,
        )
        return decode(data)

    return TtlCache(config.realtime_ttl_seconds, load)


def create_realtime_feeds(config: Config) -> CachedRealtimeFeeds:
    """Build the three cached realtime feeds for ``config``."""
    return CachedRealtimeFeeds(
        _vehicles=_feed_cache(config, config.vehicle_positions_url, decode_vehicle_positions),
        _trips=_feed_cache(config, config.trip_updates_url, decode_trip_updates),
        _alerts=_feed_cache(config, config.alerts_url, decode_alerts),
    )

File: src/realtime.ts
/**
 * GTFS-Realtime feed access: vehicle positions, trip updates, service alerts.
 *
 * Feed quirks observed in the live CATS data, which the shapes below reflect:
 *
 * - `VehiclePosition.stop_id` and `current_stop_sequence` do not correspond to
 *   any stop or sequence in the published static schedule (0% of stop ids
 *   resolved, and sequences exceed the trip's own stop count). They are
 *   therefore never surfaced as stop references; next-stop information comes
 *   from the TripUpdates feed, whose stop ids resolve completely.
 * - `StopTimeEvent.delay` is never populated, but `time` and `scheduled_time`
 *   both are, so schedule deviation is computed from those.
 */

import GtfsRealtimeBindings from 'gtfs-realtime-bindings';
import type { Config } from './config.ts';
import { fetchBinary } from './http.ts';
import { TtlCache, type Cached } from './cache.ts';

const { FeedMessage } = GtfsRealtimeBindings.transit_realtime;

export interface VehiclePosition {
  /** Agency vehicle number as shown on the bus or train, e.g. `2301`. */
  readonly vehicleLabel: string | undefined;
  readonly vehicleId: string | undefined;
  readonly entityId: string;
  readonly tripId: string | undefined;
  readonly routeId: string | undefined;
  readonly latitude: number;
  readonly longitude: number;
  readonly bearingDegrees: number | undefined;
  readonly speedMetersPerSecond: number | undefined;
  readonly occupancyStatus: string | undefined;
  readonly currentStatus: string | undefined;
  /** Epoch seconds the position was reported. */
  readonly timestamp: number | undefined;
}

export interface StopTimeUpdate {
  readonly stopId: string | undefined;
  readonly stopSequence: number | undefined;
  /** Predicted arrival, epoch seconds. */
  readonly arrivalTime: number | undefined;
  readonly scheduledArrivalTime: number | undefined;
  readonly departureTime: number | undefined;
  readonly scheduleRelationship: string | undefined;
}

export interface TripUpdate {
  readonly tripId: string | undefined;
  readonly routeId: string | undefined;
  readonly vehicleLabel: string | undefined;
  readonly vehicleId: string | undefined;
  readonly stopTimeUpdates: readonly StopTimeUpdate[];
}

export interface ServiceAlert {
  readonly headerText: string | undefined;
  readonly descriptionText: string | undefined;
  readonly cause: string | undefined;
  readonly effect: string | undefined;
  readonly severityLevel: string | undefined;
  readonly informedRouteIds: readonly string[];
  readonly informedStopIds: readonly string[];
}

/** Protobuf 64-bit fields decode as `string` under `longs: String`. */
function toEpochSeconds(value: unknown): number | undefined {
  if (value === undefined || value === null) return undefined;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : undefined;
}

function toFiniteNumber(value: unknown): number | undefined {
  if (typeof value !== 'number' || !Number.isFinite(value)) return undefined;
  return value;
}

function toText(translated: unknown): string | undefined {
  if (typeof translated !== 'object' || translated === null) return undefined;
  const translations = (translated as { translation?: unknown }).translation;
  if (!Array.isArray(translations)) return undefined;
  const english = translations.find(
    (entry) => typeof entry?.language === 'string' && entry.language.startsWith('en'),
  );
  const chosen = english ?? translations[0];
  const text = (chosen as { text?: unknown } | undefined)?.text;
  return typeof text === 'string' && text.length > 0 ? text : undefined;
}

/**
 * Decodes a GTFS-Realtime protobuf into plain objects.
 *
 * Feed payloads are untrusted input: every field is re-checked here rather
 * than assumed present, because the protobuf schema makes almost everything
 * optional and this agency omits several fields entirely.
 */
function decodeFeed(bytes: Uint8Array): { header: unknown; entities: readonly unknown[] } {
  const message = FeedMessage.decode(bytes);
  const object = FeedMessage.toObject(message, {
    longs: String,
    enums: String,
    defaults: false,
  }) as { header?: unknown; entity?: unknown };
  const entities = Array.isArray(object.entity) ? object.entity : [];
  return { header: object.header, entities };
}

export function decodeVehiclePositions(bytes: Uint8Array): VehiclePosition[] {
  const { entities } = decodeFeed(bytes);
  const vehicles: VehiclePosition[] = [];

  for (const entity of entities) {
    const record = entity as { id?: unknown; vehicle?: Record<string, unknown> };
    const vehicle = record.vehicle;
    if (vehicle === undefined) continue;

    const position = vehicle['position'] as Record<string, unknown> | undefined;
    const latitude = toFiniteNumber(position?.['latitude']);
    const longitude = toFiniteNumber(position?.['longitude']);
    // Without coordinates the entry cannot answer a location question.
    if (latitude === undefined || longitude === undefined) continue;

    const trip = vehicle['trip'] as Record<string, unknown> | undefined;
    const descriptor = vehicle['vehicle'] as Record<string, unknown> | undefined;

    vehicles.push({
      entityId: typeof record.id === 'string' ? record.id : '',
      vehicleLabel: typeof descriptor?.['label'] === 'string' ? descriptor['label'] : undefined,
      vehicleId: typeof descriptor?.['id'] === 'string' ? descriptor['id'] : undefined,
      tripId: typeof trip?.['tripId'] === 'string' ? trip['tripId'] : undefined,
      routeId: typeof trip?.['routeId'] === 'string' ? trip['routeId'] : undefined,
      latitude,
      longitude,
      bearingDegrees: toFiniteNumber(position?.['bearing']),
      speedMetersPerSecond: toFiniteNumber(position?.['speed']),
      occupancyStatus:
        typeof vehicle['occupancyStatus'] === 'string' ? vehicle['occupancyStatus'] : undefined,
      currentStatus:
        typeof vehicle['currentStatus'] === 'string' ? vehicle['currentStatus'] : undefined,
      timestamp: toEpochSeconds(vehicle['timestamp']),
    });
  }
  return vehicles;
}

export function decodeTripUpdates(bytes: Uint8Array): TripUpdate[] {
  const { entities } = decodeFeed(bytes);
  const updates: TripUpdate[] = [];

  for (const entity of entities) {
    const update = (entity as { tripUpdate?: Record<string, unknown> }).tripUpdate;
    if (update === undefined) continue;

    const trip = update['trip'] as Record<string, unknown> | undefined;
    const descriptor = update['vehicle'] as Record<string, unknown> | undefined;
    const rawStopTimes = update['stopTimeUpdate'];
    const stopTimeUpdates: StopTimeUpdate[] = [];

    if (Array.isArray(rawStopTimes)) {
      for (const raw of rawStopTimes) {
        const stopTime = raw as Record<string, unknown>;
        const arrival = stopTime['arrival'] as Record<string, unknown> | undefined;
        const departure = stopTime['departure'] as Record<string, unknown> | undefined;
        stopTimeUpdates.push({
          stopId: typeof stopTime['stopId'] === 'string' ? stopTime['stopId'] : undefined,
          stopSequence: toFiniteNumber(stopTime['stopSequence']),
          arrivalTime: toEpochSeconds(arrival?.['time']),
          scheduledArrivalTime: toEpochSeconds(arrival?.['scheduledTime']),
          departureTime: toEpochSeconds(departure?.['time']),
          scheduleRelationship:
            typeof stopTime['scheduleRelationship'] === 'string'
              ? stopTime['scheduleRelationship']
              : undefined,
        });
      }
    }

    updates.push({
      tripId: typeof trip?.['tripId'] === 'string' ? trip['tripId'] : undefined,
      routeId: typeof trip?.['routeId'] === 'string' ? trip['routeId'] : undefined,
      vehicleLabel: typeof descriptor?.['label'] === 'string' ? descriptor['label'] : undefined,
      vehicleId: typeof descriptor?.['id'] === 'string' ? descriptor['id'] : undefined,
      stopTimeUpdates,
    });
  }
  return updates;
}

export function decodeAlerts(bytes: Uint8Array): ServiceAlert[] {
  const { entities } = decodeFeed(bytes);
  const alerts: ServiceAlert[] = [];

  for (const entity of entities) {
    const alert = (entity as { alert?: Record<string, unknown> }).alert;
    if (alert === undefined) continue;

    const routeIds = new Set<string>();
    const stopIds = new Set<string>();
    const informed = alert['informedEntity'];
    if (Array.isArray(informed)) {
      for (const raw of informed) {
        const selector = raw as Record<string, unknown>;
        if (typeof selector['routeId'] === 'string') routeIds.add(selector['routeId']);
        if (typeof selector['stopId'] === 'string') stopIds.add(selector['stopId']);
      }
    }

    alerts.push({
      headerText: toText(alert['headerText']),
      descriptionText: toText(alert['descriptionText']),
      cause: typeof alert['cause'] === 'string' ? alert['cause'] : undefined,
      effect: typeof alert['effect'] === 'string' ? alert['effect'] : undefined,
      severityLevel:
        typeof alert['severityLevel'] === 'string' ? alert['severityLevel'] : undefined,
      informedRouteIds: [...routeIds],
      informedStopIds: [...stopIds],
    });
  }
  return alerts;
}

export interface RealtimeFeeds {
  vehiclePositions(): Promise<Cached<VehiclePosition[]>>;
  tripUpdates(): Promise<Cached<TripUpdate[]>>;
  alerts(): Promise<Cached<ServiceAlert[]>>;
}

export function createRealtimeFeeds(config: Config): RealtimeFeeds {
  const build = <T>(url: string, decode: (bytes: Uint8Array) => T): TtlCache<T> =>
    new TtlCache<T>(config.realtimeTtlMs, async () => {
      const bytes = await fetchBinary(url, {
        timeoutMs: config.requestTimeoutMs,
        maxBytes: config.maxFeedBytes,
      });
      return decode(bytes);
    });

  const vehicles = build(config.vehiclePositionsUrl, decodeVehiclePositions);
  const trips = build(config.tripUpdatesUrl, decodeTripUpdates);
  const alerts = build(config.alertsUrl, decodeAlerts);

  return {
    vehiclePositions: () => vehicles.get(),
    tripUpdates: () => trips.get(),
    alerts: () => alerts.get(),
  };
}
